compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -915,6 +912,61 @@ impl<'a> Parser<'a> {
         Ok(PatKind::TupleStruct(qself, path, fields))
     }
 
+    /// Parses `box pat`
+    fn parse_pat_box(&mut self) -> PResult<'a, PatKind> {
+        let box_kw = self.token.ident().unwrap().0;
```

Comment:
We need the whole keyword, not just its span. Do you mean to create a new `Ident` with this span? Why do that?

I can use `?` but I don't think it makes sense.

Reply:
> I can use `?` but I don't think it makes sense.

There's a bunch of parser code that intentionally calls functions that return `PResult<'_, T>` for its error value -- calling this incorrectly (i.e. when the current token isn't `box`) shouldn't ICE, imo.

Reply:
> Do you mean to create a new Ident with this span? Why do that?

To avoid an ICE, and/or just keep the "regular" parse path simpler. The `box_kw` ident is only needed on the recovery, so it should be equivalently as easy to just create it as needed with `kw::Box` symbol and the span we have easy access to.

Reply:
What about
```rust
let box_kw = self.token.ident().ok_or_else(|| self.unexpected())?.0;
```

Reply:
Why not just avoid doing the check twice at all? A lot of parser functions expect initial tokens to be already parsed. Why not just change `check` back to `eat` in the caller, and start off your `parse_box_pat` like `parse_pat_ident_mut` does: https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_parse/parser/pat.rs.html#574-575

I don't think you should be going thru all of this (`unwrap` or mapping the error) _just_ to avoid `Ident::new` on one line in recovery code.

Reply:
Pushing soon.